src/Channels.Networking.Sockets/SocketConnection.cs

```diff
@@ -0,0 +1,304 @@
+﻿using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Channels.Networking.Sockets
+{
+    public class SocketConnection : IChannel
+    {
+        private Socket _socket;
+        private Channel _input, _output;
+        private ChannelFactory _channelFactory;
+        private readonly bool _ownsChannelFactory;
+
+        static readonly EventHandler<SocketAsyncEventArgs> _asyncCompleted = OnAsyncCompleted;
+
+        public IReadableChannel Input => _input ?? CreateInput();
+        public IWritableChannel Output => _output ?? CreateOutput();
+
+        private Channel CreateInput()
+        {
+            _input = ChannelFactory.CreateChannel();
+            ProcessReads();
+            return _input;
+        }
+
+        private Channel CreateOutput()
+        {
+            _output = ChannelFactory.CreateChannel();
+            ProcessWrites();
+            return _output;
+        }
+
+        private Socket Socket => _socket;
+        private ChannelFactory ChannelFactory => _channelFactory;
+
+        internal SocketConnection(Socket socket, ChannelFactory channelFactory)
```

Comment:
I guess it comes down to what the minimal API is for creating a client socket - ConnectAsync etc. We can force the external code to declare it, but seems a little ugly.